src/hotspot/share/gc/g1/g1Allocator.cpp

```diff
@@ -452,26 +452,29 @@ HeapWord* G1ArchiveAllocator::archive_mem_allocate(size_t word_size) {
          PTR_FORMAT " <= " PTR_FORMAT " <= " PTR_FORMAT,
          p2i(_bottom), p2i(old_top), p2i(_max));
 
-  // Allocate the next word_size words in the current allocation chunk.
-  // If allocation would cross the _max boundary, insert a filler and begin
-  // at the base of the next min_region_size'd chunk. Also advance to the next
-  // chunk if we don't yet cross the boundary, but the remainder would be too
-  // small to fill.
-  HeapWord* new_top = old_top + word_size;
-  size_t remainder = pointer_delta(_max, new_top);
-  if ((new_top > _max) ||
-      ((new_top < _max) && (remainder < CollectedHeap::min_fill_size()))) {
+  // Try to allocate word_size in the current allocation chunk. Two cases
+  // require special treatment:
+  // 1. no enough space for word_size
+  // 2. after allocating word_size, there's non-zero space left, but too small for the minimal filler
+  // In both cases, we retire the current chunk and move on to the next one.
+  size_t free_words = pointer_delta(_max, old_top);
+  if (free_words < word_size ||
+      ((free_words - word_size != 0) && (free_words - word_size < CollectedHeap::min_fill_size()))) {
+    // Retiring the current chunk
     if (old_top != _max) {
-      size_t fill_size = pointer_delta(_max, old_top);
+      // Non-zero space; need to insert the filler
+      size_t fill_size = free_words;
       CollectedHeap::fill_with_object(old_top, fill_size);
       _summary_bytes_used += fill_size * HeapWordSize;
     }
+    // Set the current chunk as "full"
     _allocation_region->set_top(_max);
-    old_top = _bottom = _max;
 
     // Check if we've just used up the last min_region_size'd chunk
     // in the current region, and if so, allocate a new one.
-    if (_bottom != _allocation_region->end()) {
+    if (_max != _allocation_region->end()) {
+      // Shift to the next chunk
+      old_top = _bottom = _max;
```

Comment:
Why is this line being moved?  It seems otherwise unrelated to the CR change.  With this change, if `alloc_new_region` fails then `_bottom` is no longer updated; is that okay?  I spent a little time investigating, but haven't yet found anything convincing either way.

Reply:
> Why is this line being moved?

Two reasons:

1. To make it more symmetric in updating `old_top`, `_bottom` and `_max`.

```
if (_max != _allocation_region->end()) {
  ... // updating the triplet in one way
} else {
  ... // updating them in another way
}
```

2. In general, the range of `[_bottom, _top]` marks the allocatable space, and `_top` fits in the middle. In the original code, when `alloc_new_region` fails, `_bottom == _max`, which leaves zero allocatable space. With this PR, `[_bottom, _top]` is still a valid allocatable space, but full.

> is that okay?

When `alloc_new_region` fails, `G1ArchiveAllocator::archive_mem_allocate` returns NULL. Callers seem to expect this method to *not* return NULL in the normal case; they either terminate the VM (`HeapShared::archive_heap_object`) or directly treat the return value as non-null (`HeapShared::copy_roots()` and `G1ArchiveAllocator::complete_archive`). Therefore, both the existing code and the PR are fine.

BTW, it seems that `_bottom` is only used for verification purpose; maybe remove it or make this more explicit.

Reply:
Thanks for the explanation.